src/views/Package/PackageState.tsx

```diff
@@ -50,6 +76,165 @@ export const usePackageState = () => {
+  }: { scope?: string; language: string; name: string; version: string }
+): { readme: string; apiReference: Types; menuItems: MenuItem[] } => {
+  const nameSegment = scope ? `${scope}/${name}` : `${name}`;
+  const basePath = `/packages/${nameSegment}/v/${version}`;
+  const langQuery = `?${QUERY_PARAMS.LANGUAGE}=${language}`;
+  const separator =
+    '\n# API Reference <span data-heading-title="API Reference" data-heading-id="api-reference"></span>\n';
+
+  // split into readme and api reference
+  const segments = input.split(separator);
+
+  // Take the last chunk after the separator
+  // segments.pop() always returns when length > 1;
+  const apiReferenceStr = segments.length > 1 ? segments.pop()! : "";
+
+  // Rejoin all the previous chunks in case the readme has the same Separator
+  const readmeStr = segments.join(separator);
+
+  //split each on headers
+  const apiReferenceSplit = splitOnHeaders(apiReferenceStr, 3);
+  const readmeSplit = splitOnHeaders(readmeStr, 6);
+
+  // Add back api reference title for use as menu item
+  const apiReferenceParsed = [separator.trim(), ...(apiReferenceSplit ?? [])];
+
+  const baseReadmePath = `${basePath}${langQuery}`;
+  const readmeMenuItems = [
+    {
+      level: 1,
+      id: "Readme",
+      title: "Readme",
+      path: baseReadmePath,
+      children: readmeSplit.reduce((accum: MenuItem[], str: string) => {
+        if (str.startsWith("#")) {
+          const { id, title } = getHeaderAttributes(str);
+          const level = str.match(/(#)/gm)?.length ?? 1;
+          const menuItem = {
+            level,
+            id,
+            title,
+            // root package path plus hash for header on readme item
+            path: `${baseReadmePath}#${id}`,
+            children: [],
+          };
+          return appendMenuItem(accum, menuItem);
+        }
+        return accum;
+      }, []),
+    },
+  ];
+
+  let menuItems: MenuItem[] = [...readmeMenuItems];
+  const types: Types = {};
+
+  const getApiPath = (id: string) => `${basePath}/api/${id}${langQuery}`;
+  let prevType: { id: string; title: string };
+  apiReferenceParsed?.forEach((str) => {
+    // TODO get attributes off of embedded html
+    const isHeader = str.match(/(^#{1,3}[^#].*)/gm);
+    if (isHeader?.length) {
+      const { id, title } = getHeaderAttributes(str);
+      const level = str.match(/(#)/gm)?.length ?? 1;
+
+      // root package path plus type id segment
+      // only level 3 headers are types in api reference
+      const path = level === 3 ? getApiPath(id) : undefined;
+      const menuItem = {
+        level,
+        id,
+        title,
+        children: [],
+        ...(path ? { path } : {}),
+      };
+
+      menuItems = appendMenuItem(menuItems, menuItem);
+      prevType = { id, title };
+    } else {
+      types[prevType.id] = { title: prevType.title, content: str };
+    }
+  });
+
+  return {
+    readme: segments.join(separator),
+    apiReference: types,
+    menuItems,
+  };
+};
+
```

Comment:
I'm not sure this is necessarily page-level state or needs to be. Do you think it would make sense to encapsulate this logic within a `PackageDocsState.tsx` or even just a util that is used by the docs component?

Reply:
idk I think `menuItems` + `apiReference` + `readme` is basically page level for the package page. It's not state, it's all derived, but if the md doc itself is page level I'd argue that all of this is too. We could definitely compute this within the component but thinking forward, if this is the format we ideally would want the backend to send to us, I think doing this type of normalization right after receiving the response is the correct place.

In practice  we can put it wherever, but creating another context seems unnecessarily contrived when all of this processing is done from the result of the md document and the document itself is basically no longer useful. I'd just say make a utils file right next to this that exports `parseMarkdownStructure` but that's all I'd want to separate off the top of my head.

Reply:
Extracted into a `util.ts` file for now. Down to revisit restructuring contexts/hooks here in a follow up if you want to discuss it more.